platform/pulp/platform/tasks/task_system.py

```diff
@@ -1,4 +1,548 @@
+
+    :returns:          The Worker instance that has no reserved_resource
+                       entries associated with it.
+    :rtype:            pulp.platform.model.Worker
+    """
+    free_workers = [Worker.objects.annotate(Count('reservations')).filter(reservations__count=0)]
+    if len(free_workers) == 0:
+        raise ReservedResource.DoesNotExist()
+    random.shuffle(free_workers)
+    return free_workers[0]
+
+
+def _delete_worker(name, normal_shutdown=False):
+    """
+    Delete the Worker with name from the database, cancel any associated tasks and reservations
+
+    If the worker shutdown normally, no message is logged, otherwise an error level message is
+    logged. Default is to assume the worker did not shut down normally.
+
+    Any resource reservations associated with this worker are cleaned up by this function.
+
+    Any tasks associated with this worker are explicitly canceled.
+
+    :param name:            The name of the worker you wish to delete.
+    :type  name:            basestring
+    :param normal_shutdown: True if the worker shutdown normally, False otherwise.  Defaults to
+                            False.
+    :type normal_shutdown:  bool
+    """
+    if normal_shutdown is False:
+        msg = _('The worker named %(name)s is missing. Canceling the tasks in its queue.')
+        msg = msg % {'name': name}
+        _logger.error(msg)
+
+    # Cancel all of the tasks that were assigned to this worker's queue
+    for task_status in Worker.objects.tasks.filter(state__in=constants.CALL_INCOMPLETE_STATES):
+        cancel(task_status.id)
+
+    # Delete the worker document
+    Worker.objects.get(name=name).delete()
+
+    # If the worker is a resource manager, we also need to delete the associated lock
+    if name.startswith(RESOURCE_MANAGER_WORKER_NAME):
+        ResourceManagerLock.objects.filter(name=name).delete()
+
+
+@task(base=PulpTask)
+def _release_resource(task_id):
+    """
+    Do not queue this task yourself. It will be used automatically when your task is dispatched by
+    the _queue_reserved_task task.
+
+    When a resource-reserving task is complete, this method releases the resource by removing the
+    ReservedResource object by UUID.
+
+    :param task_id: The UUID of the task that requested the reservation
+    :type  task_id: basestring
+    """
+    try:
+        TaskStatus.objects.get(id=task_id, state=constants.CALL_RUNNING_STATE)
+    except TaskStatus.DoesNotExist:
+        pass
+    else:
+        new_task = Task()
+        msg = _('The task status %(task_id)s exited immediately for some reason. Marking as '
+                'errored. Check the logs for more details')
+        runtime_exception = RuntimeError(msg % {'task_id': task_id})
+
+        class MyEinfo(object):
+            traceback = None
+
+        new_task.on_failure(runtime_exception, task_id, (), {}, MyEinfo)
+
+    ReservedResource.objects.filter(task__id=task_id).delete()
```

Comment:
so does it mean that if using double underscore id you are accessing id of the task object?( and you can do that because task field on ReservedResource model is one-to-one field?)